clang/include/clang/Sema/Sema.h

```diff
@@ -13925,20 +13925,32 @@ class Sema final {
                                    SourceLocation BuiltinLoc,
                                    SourceLocation RParenLoc);
 
-  template <typename AttrTy>
-  bool isTypeDecoratedWithDeclAttribute(QualType Ty) {
+  template <typename AttrTy> const AttrTy *getAttrTypeDecl(QualType Ty) {
     const CXXRecordDecl *RecTy = Ty->getAsCXXRecordDecl();
     if (!RecTy)
-      return false;
+      return nullptr;
 
-    if (RecTy->hasAttr<AttrTy>())
-      return true;
+    if (const auto *A = RecTy->getAttr<AttrTy>())
+      return A;
 
     if (auto *CTSD = dyn_cast<ClassTemplateSpecializationDecl>(RecTy)) {
       ClassTemplateDecl *Template = CTSD->getSpecializedTemplate();
-      if (CXXRecordDecl *RD = Template->getTemplatedDecl())
-        return RD->hasAttr<AttrTy>();
+      CXXRecordDecl *RD = Template->getTemplatedDecl();
+      return RD ? RD->getAttr<AttrTy>() : nullptr;
     }
+    return nullptr;
+  }
+
+  template <typename AttrTy>
+  bool isTypeDecoratedWithDeclAttribute(QualType Ty) {
+    if (const auto *A = getAttrTypeDecl<AttrTy>(Ty))
+      return true;
+    return false;
+  }
+
+  bool isSyclType(QualType Ty, SYCLTypeAttr::SYCLType TypeName) {
```

Comment:
Just some thoughts...
I recently had some experience that involved looking into how the types are formed in clang ast, so I got under impression that functionality connected to types including the ones like `isItTheType()` is usually owned by ASTContext rather than Sema (maybe I'm wrong and @AaronBallman and @erichkeane could opt in here).
Since Sema and Codegen usually both have access to ASTContext, we could unify the way the device_global type is detected.

Reply:
Just a drive-by, I think @Fznamznon 's thoughts on the 'type identification' functions is correct here.

Reply:
The identification in this case is essentially just detection of `sycl_type` attribute. This new function `isSyclType` calls an existing Sema function  - `isTypeDecoratedWithDeclAttribute`,  which has just been renamed and refactored to return the attribute itself instead of a bool value. 

I do not know how type identification normally takes place in ASTContext but I would assume it is not by checking attributes, which is why I am not convinced it belongs there.  I can move all of this to ASTContext but that moves/changes existing code